### ISSUE TITLE
Default typing does not work with `writerFor(Object.class)`

### ISSUE THREAD

**Issue by hoomanv** [FasterXML/jackson-databind#1093]
I have enabled default typing for OBJECT_AND_NON_CONCRETE.
I expect that the following should include type information for rootObject itself.
objectMapper.writerFor(Object.class).write(rootObject)
But it doesn't. It only includes type information for fields (if they are object or non-concrete)
Because writerFor(Object.class) is actually similar to not setting any root type.
So what can I do to include type information for the root object ?
One workaround is to wrap it in an Object[], but I don't like extra array tokens in the final json

**Comment by cowtowncoder** [comment]
This is tricky one, but only because type `java.lang.Object` is somewhat overloaded. In many contexts it is taken to mean "use natural types", that is, mapping JSON into combination of `Map`s, `List`s, `String`s and wrappers. So this may be why typing does not work.

I think you are right in that it should work, so above is just speculation of what might be going wrong.

A simple unit test would help diagnosing here, although I should be able to compose one if you don't have one. There are existing tests for default typing just not for this combination of calls.

**Comment by cowtowncoder** [comment]
Actually I do think that I need the unit test here.

This may also be related to long-standing ideas of needing distinction between serialization type (for which existing functionality is designed; use given type instead of real runtime), and base type used for polymorphic handling. The two are tightly bound, but they shouldn't be: in your case, you'd want to use this as base type, but not for serialization of properties.

**Comment by hoomanv** [comment]
serialization-type defines the perspective to take when looking at an object. It controls what properties will or will not be included depending on how we look at the object. It is usually determined by the runtime type of the object but we can overwrite it using writerFor method.

What base-type is going to tell us about?
Would it just tell the type of the object as if it was declared as a field?
Sounds like default-typing is actually taking this type into account in order to decide whether to include the type information or not.

So my idea is that writerFor(Class) is not actually a good place to fit it in. We need a way to distinguish between the two concepts.
May be we need another method like writerForDefaultType(). I'm not sure

**Comment by cowtowncoder** [comment]
Base type tells, during serialization, what is the expected type given for `readValue()`, basically. Or, technically, what `@JsonTypeInfo` definition is to be used, so, class for which it is used (usually directly, but possibly via base class). Base type is the starting point for polymorphic type resolution; but actual properties included may only exist in actual subtype that was serialized.

So yes, it would essentially tell declared type of the field, since that is typically the main piece of contextual information available during deserialization.

And yes, you are following with the ideas I have had: separate base type. I have almost implemented it twice already, but both times started doubting whether this was needed, and what would be the canonical use case...

What would be great would, I think, be a unit test that shows the problem with existing usage; case where `writerFor(...)` does not work. So far I have thought that this mostly applies to container types (like `List` or `Map`); if a POJO type could be demonstrated that would be great.

**Comment by hoomanv** [comment]
@cowtowncoder here you go

```java
import static com.fasterxml.jackson.annotation.JsonTypeInfo.As.WRAPPER_ARRAY;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import static com.fasterxml.jackson.databind.ObjectMapper.DefaultTyping.OBJECT_AND_NON_CONCRETE;
import java.util.Objects;
import static org.junit.Assert.*;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class TestDefaultTyping
{
    private static final Logger LOGGER = LoggerFactory.getLogger(TestDefaultTyping.class);

    @Test
    public void testPojo() throws Exception
    {
        ObjectMapper mapper = new ObjectMapper();
        mapper.enableDefaultTyping(OBJECT_AND_NON_CONCRETE, WRAPPER_ARRAY);

        Object expected = new Foo(new Foo("bar"));
        String json = mapper.writerFor(Object.class).writeValueAsString(expected);
        LOGGER.debug("json = {}", json);

        Object actual = mapper.readerFor(Object.class).readValue(json); //FIXME it fails
        assertEquals(expected, actual);

        try
        {
            mapper.readerFor(Foo.class).readValue(json);
            fail(); //FIXME it does not fail
        } catch (JsonMappingException ex)
        {
            LOGGER.debug("Must Fail", ex);
        }
    }

    @Test
    public void testObjectArray() throws Exception
    {
        ObjectMapper mapper = new ObjectMapper();
        mapper.enableDefaultTyping(OBJECT_AND_NON_CONCRETE, WRAPPER_ARRAY);

        Object[] expected = new Foo[]
        {
            new Foo(new Foo("bar"))
        };
        String json = mapper.writerFor(Object[].class).writeValueAsString(expected);
        LOGGER.debug("json = {}", json);

        Object[] actual = mapper.readerFor(Object[].class).readValue(json);
        assertArrayEquals(expected, actual);

        try
        {
            mapper.readerFor(Foo[].class).readValue(json);
            fail();
        } catch (JsonMappingException ex)
        {
            LOGGER.debug("Must Fail", ex);
        }
    }

    public static class Foo
    {
        private Object value;

        public Foo()
        {
        }

        public Foo(Object value)
        {
            this.value = value;
        }

        public Object getValue()
        {
            return value;
        }

        public void setValue(Object value)
        {
            this.value = value;
        }

        @Override
        public int hashCode()
        {
            int hash = 7;
            hash = 53 * hash + Objects.hashCode(this.value);
            return hash;
        }

        @Override
        public boolean equals(Object obj)
        {
            if (obj == null)
            {
                return false;
            }
            if (getClass() != obj.getClass())
            {
                return false;
            }
            final Foo other = (Foo) obj;
            return Objects.equals(this.value, other.value);
        }

    }
}
```

**Comment by cowtowncoder** [comment]
I think #789 might be duplicate here, closing as dup.

**Comment by cowtowncoder** [comment]
Actually I think I should keep this issue open, being specific, and close the other issue since it is more general and has no related issues as of yet.

**Comment by cowtowncoder** [comment]
@hoomanv Belated thank you for reporting this; will finally be fixed in 2.10.0, to be released with in 2 weeks (and was included in pre-release 3)
